### DISCUSSION TITLE
ASAR error when used with Electron

### DISCUSSION THREAD

**Discussion by AkshayPathak** [prisma/prisma#5200] [Q&A] [answered]
Hi there,

Thank you so much for all the hard work that goes into a project like this.

I am getting an error `Error: spawn` when I use it with Electron. This is the full error:

```
Error: spawn
**********out\win-unpacked\resources\app.asar\query-engine-windows.exe ENOENT
```

I have verified the `query-engine-windows.exe` is inside `app.asar`.

I found this issue #4356 but I am not sure how I can fix this particular issue.

Thank you for the help.

**Comment by sdnts** [comment, marked as answer]
Hi @AkshayPathak! We ship the query engine binary with Studio, so I can definitely help you with this.

First things first, you want to make sure that you generate Prisma Client against your schema BEFORE packaging your app. It is possible to generate Client on the fly (like we do in Studio), but that process is a lot more involved. In your case, I'm assuming you do not need to support multiple schemas, so it'll probably be easier to generate Client up front.

Now that you have everything ready to be pacjaged, you should know that executables inside an ASAR archive cannot work. Normally, `electron-builder` (if you use it) automatically "unpacks" any executables to an `app.asar.unpacked` folder. Does this not happen for you? In order to `spawn` the Query Engine binary, it has to be outside the ASAR archive, since ASAR archives are read-only, and it is not possible to "peek" inside them. You should look in this location to possibly find the query engine: `<your-electron-repo>/win-unpacked/resources/app.asar.unpacked/node_modules/@prisma/engines/query-engine-windows.exe`.

Once you do that, you'll have to tell Prisma Client where this query engine binary (QE for short) was relocated to. Normally, Prisma Client can automatically locate the QE, but in this specific case, since you (electron-builder) moved it out, you need to tell Client where it is. In it's constructor, you can tell Prisma Client where to locate the QE this way:

```js
let prisma = new PrismaClient({
   __internal: {
    engine: {
      binaryPath: "/path/to/binary"
    }
  }
})
```
In your case, if the QE gets unpacked by `electron-builder` automatically, the path *should* be:
```js
import path from 'path'
import { app } from 'electron'

const qePath = path.join(
  app.getPath().replace('app.asar', 'app.asar.unpacked'),
  'node_modules/@prisma/engines/query-engine-windows.exe'
)
```
Again, you should also be able to locate it in Explorer in `<your-electron-repo>/win-unpacked/resources/app.asar.unpacked/node_modules/@prisma/engines/query-engine-windows.exe`.
(Note that I'm only considering Windows here, if you support macOS and Linux, the names of the binaries are slightly different)

If you have any other questions, or need more help, let me know!

**Reply by AkshayPathak** [reply]
Hi @madebysid! Thank you very much for the very-detailed response. I am new to Electron and I am indeed using Electron Builder. The `app.asar.unpacked` folder was not generating but I was able to manually get it to unpack the QE with this Electron Builder configuration:

```
"win": {
    "target": "nsis",
    "asar": true,
    "asarUnpack": [
        "**/query-engine-windows.exe"
    ]
},
``` 

Thank you also for sharing how I can set the path to the binary when creating the client. I think I am all good on this front.

**Reply by vladejs** [reply]
Hi, I managed to point Prisma to the engine on the `app.asar.unpacked` folder, but I'm getting this error:

```
(node:16436) UnhandledPromiseRejectionWarning: Error: spawn C:\Users\User\OneDrive\Desktop\projects\ddd\release\win-unpacked\resources\app.asar.unpacked\node_modules\@prisma\engines\query-engine-windows.exe ENOENT
```

If I run that path from the command line, it gets executed:
```
> C:\Users\User\OneDrive\Desktop\projects\ddd\release\win-unpacked\resources\app.asar.unpacked\node_modules\@prisma\engines\query-engine-windows.exe

{"is_panic":false,"message":"Datamodel should be provided either as path or base64-encoded string.","backtrace":"   0: <unknown>\n   1: <unknown>\n   2: <unknown>\n   3: <unknown>\n   4: <unknown>\n   5: <unknown>\n   6: <unknown>\n   7: <unknown>\n   8: <unknown>\n   9: <unknown>\n  10: <unknown>\n  11: <unknown>\n  12: <unknown>\n  13: <unknown>\n  14: <unknown>\n  15: <unknown>\n  16: <unknown>\n  17: <unknown>\n  18: <unknown>\n  19: BaseThreadInitThunk\n  20: RtlUserThreadStart\n"}
```

Electron 12
Node 14.17
Electron-builder 22.11.7

**Reply by vladejs** [reply]
There is also a [SO question](https://stackoverflow.com/questions/67962157/prisma-client-not-finding-query-engine-after-electron-packages-everything)

**Reply by sdnts** [reply]
Hey I'm just seeing this, is this also happening on the latest versions of everything? (Prisma is at 2.30.3)

**Comment by MitchellMonaghan** [comment]
I'm struggling to get this working. I fixed the same  error with the suggested asarUnpack but now I get "Could not open datamodel file "C:\projects\myApp\dist\electron\Packaged\win-unpacked\resources\app.asar\schema.prisma"

I didn't point it to the binary but it seems to be able to find it anyway? I tried unpacking the Schema file but the engine still looks in the same spot. There doesn't seem to be a way to point the query engine at a different schema location.

https://github.com/prisma/prisma/issues/2746

Maybe I don't really understand what's happening.

**Reply by MitchellMonaghan** [reply]
Also mine uses a modified prisma generate output path b/c I can't seem to get webpack to include the node_modules/.prisma 

Instead I'm using copy webpack plugin

**Comment by sdnts** [comment]
Hey @MitchellMonaghan, do you happen to have a repository I can look at? There's a bunch of moving pieces, so looking at code should help me figure out what's up.

IIUC, you're able to package the query engine binary with your app (using the copy webpack plugin), but on launch, the query engine expects to find your schema at `C:\projects\myApp\dist\electron\Packaged\win-unpacked\resources\app.asar\schema.prisma`?

There are ways to override this, but they're very obscure (and very much implementation details) and may change without notice, so I'd like to make sure you avoid depending on them.

One very generic solution I can give you is to try out the new `napi` flag (https://github.com/prisma/prisma/releases/tag/2.20.0). TL;DR is that the new N-API equipped query engine does not try to read the schema from a specific path, so your problem should not exist anymore. N-API is still in preview, but we plan on stabilizing it soon.

Another generic (but incorrect) solution would be to place your schema in the location the query engine is looking for it in, and see if that works. If it does, we can start back-tracking.

But if you give me some code, I'll be able to help you better!

**Reply by annettaws** [reply]
Hi i'm stucking still on the problem of working prisma query engine windows in build state.
The project is electron + express server and prisma.
I have applied you suggestion but on building the electron app on window (with the query-engine-window unpacked) does not work.
Follow the branch [project](https://github.com/tiappendi/vanlio/tree/prelect).

**Comment by DubG1** [comment]
Hello, i have run into a similiar error after building my electron app with prisma, i am using vite and electron builder

![image](https://github.com/prisma/prisma/assets/104206714/047c7be6-1f8f-4cb9-bc92-91613a08288b)


after looking up some stuff i changed my vite config

`rollupOptions: {
                external: Object.keys('dependencies' in pkg ? pkg.dependencies : {}).filter(
                  (dependency) => dependency !== '@prisma/client' // Don't exclude @prisma/client
                ),
 `
 
 so it wouldnt exclude prisma client and that fixed some errors for me so now i "only" have this error, 
but now prisma doesnt work in development, possibly i need to change how i import/declare my client in my main index.ts depending on dev or production? it seems like it has something to do with vite
 
 
![image](https://github.com/prisma/prisma/assets/104206714/189e293c-943d-4f87-bda3-9c4bb9193a04)

if anyone needs more info i can create a repo with my config files if that helps

**Reply by kaleem-Durrani** [reply]
Did you by any chance solve this issue? 

I have the exact same configuration and the exact same error that you have in the first image. 

If you have found a solution could you please share it.

**Reply by DubG1** [reply]
ok so i had to add some stuff so everything necessary for prisma is included when building, this is my build script from my package.json

```
"build": {
    "directories": {
      "output": "release/${version}"
    },
    "appId": "com.yourAppId",
    "win": {
        "artifactName": "${productName}_${version}.${ext}",
        "target": "NSIS",
        "icon": "iconpath"
    },
    "files": [
        "dist-electron",
        "dist",
        "prisma/**/*",
        "node_modules/.prisma/**/*",
        "node_modules/@prisma/client/**/*",
        "node_modules/@prisma/engines/migration-engine*",
        "node_modules/@prisma/engines/query*",
        "node_modules/@prisma/engines/libquery*"
    ],
    "extraResources": [
        "prisma/**/*",
        "node_modules/.prisma/**/*",
        "node_modules/@prisma/client/**/*",
        "node_modules/@prisma/engines/migration-engine*",
        "node_modules/@prisma/engines/query*",
        "node_modules/@prisma/engines/libquery*"
    ]
}

**Reply by kaleem-Durrani** [reply]
Thank you, although i was not using the package.json, instead i was using electron-builder.yml for configuration. i tried applying this with the help of claude.ai, and it worked, though i have to be honest i dont understand why, or what i was doing wrong before.
This is how my electron builder.yml looks now:

```appId: com.sm-trading.app
productName: sm-trading
directories:
  buildResources: build
  output: dist
files:
  - out/**/*
  - '!**/.vscode/*'
  - '!src/*'
extraResources:
  - from: 'node_modules/.prisma/client'
    to: 'node_modules/.prisma/client'
  - from: 'node_modules/@prisma/client'
    to: 'node_modules/@prisma/client'
  - from: 'node_modules/@prisma/engines'
    to: 'node_modules/@prisma/engines'
  - from: 'prisma'
    to: 'prisma'
  - from: 'config/production.json'
    to: 'config/production.json'
asar: true
asarUnpack:
  - 'node_modules/.prisma/**/*'
  - 'node_modules/@prisma/**/*'
  - 'prisma/**/*'
win:
  executableName: sm-trading
nsis:
  artifactName: ${name}-${version}-setup.${ext}
  shortcutName: ${productName}
  uninstallDisplayName: ${productName}
  createDesktopShortcut: always
mac:
  entitlementsInherit: build/entitlements.mac.plist
  extendInfo:
    - NSCameraUsageDescription: Application requests access to the device's camera.
    - NSMicrophoneUsageDescription: Application requests access to the device's microphone.
    - NSDocumentsFolderUsageDescription: Application requests access to the user's Documents folder.
    - NSDownloadsFolderUsageDescription: Application requests access to the user's Downloads folder.
  notarize: false
dmg:
  artifactName: ${name}-${version}.${ext}
linux:
  target:
    - AppImage
    - snap
    - deb
  maintainer: Kaleemullah.com
  category: Utility
appImage:
  artifactName: ${name}-${version}.${ext}
npmRebuild: false
```

and it works perfectly.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
